Fixed issue where metadata would come with garbage

Tjoms99 opened the pull request:
Investigated the PR/issue discussed here #49. 
 
 Turns out not properly exiting a program running the PPK2 influences the next PPK2 session.  Setting up the metadata the first time will throw an exception due to the garbage data (from the previous session) that comes a long with it. Physically turning the PPK2 on and off using the switch resolves this issue, but that’s inconvenient to do every time; So I’ve added a software fix here :)
 
 To reproduce the issue and showcase the fix
 
 1. Connect the PPK2 via USB
 
 2. Open a terminal and run the example 
 
 ```
 python3 examples/example_measure_mw.py
 ```
 
 3.  Terminate the program (2x)
 ```
 ctrl + c
 ```
 4. Run the example again and you'll probably see this in the terminal
  ```
  Attempt 1/2 - Failed to get valid PPK2 metadata: Could not retrieve valid metadata from the device.
  Attempt 2/2 - Got metadata from PPK2
  Initialized Power Profiler
  ```
 
 This is where the `UnicodeDecodeError` would previously be triggered

Tjoms99 committed:
Fixed issue where unicode exception would be thrown if a previous PPK2 session were not terminated properly due to garbage data coming along with the metadata. We now expect them and try to get metadata one more time, but this time without garbage from previous session

Added example to reproduce/showcase error and fix. This examples should also just work after connecting to a PPK2 via USB

Empty buffer on ppk2 init, remove retries on metadata as it was rather treating the symptom instead of the root cause

Add retries on connection as a bad exit can cause the serial port to not be immidiatly discovered

Remove logging and raise exceptions instead. Calling logger.*(message) messes up with higher level application loggers. The application using the ppk2 lib should decide what to do with the exceptions and thereby log what they need.

Fixed borken api call in power_profiler.py

Harden API by adding types to functions

cleanup import

Tjoms99 committed:
Fixed issue where unicode exception would be thrown if a previous PPK2 session were not terminated properly due to garbage data coming along with the metadata. We now expect them and try to get metadata one more time, but this time without garbage from previous session

Tjoms99 committed:
Added example to reproduce/showcase error and fix. This examples should also just work after connecting to a PPK2 via USB

RmThrt commented:
Thanks for the fix, it helped to figure out the way my test work! Good fix for me :)

MrKevinWeiss commented:
Also tested and it works for me!

MrKevinWeiss commented:
This is on 4.2.0 and 4.2.1 fw.

mstixenberger commented:
works fine, thank you!

bobatsar commented:
Works also on my side.  @wlgrd is it possible to merge it into master?

tphillips-ict commented:
+1 Can we merge this? Running into the same issue.

Is this project still actively maintained?

MrKevinWeiss commented on src/ppk2_api/ppk2_api.py:
```diff
@@ -229,17 +241,40 @@ def list_devices():
-        metadata = self._read_metadata()
-        ret = self._parse_metadata(metadata)
-        return ret
+    def get_modifiers(self, retries=2):
```
```suggestion
    def get_modifiers(self, retries=4):
```

I have seen this occurring in our devices, maybe a bit of extra time would help?

MrKevinWeiss commented on src/ppk2_api/ppk2_api.py:
```diff
@@ -229,17 +241,40 @@ def list_devices():
+
+        
+        print("Failed to get modifiers after multiple attempts.")
+        return None
```
```suggestion
        for attempt in range(1, retries + 1):
            # Send command to request metadata
            self._write_serial((PPK2_Command.GET_META_DATA, ))
            try:
                metadata = self._read_metadata()    
                ret = self._parse_metadata(metadata)
                logging.info(f"Attempt {attempt}/{retries} - Got metadata from PPK2")

                return ret
            except ValueError as e:
                logging.warning(f"Attempt {attempt}/{retries} - Failed to get valid PPK2 metadata: {e}")
                # Wait a bit in case device was previously running
                time.sleep(0.5)
                # Clear anything in the buffer
                self.ser.read(self.ser.in_waiting)
                # If this wasn't the last attempt, we try again by sending GET_META_DATA again.
        raise IOError("Failed to get modifiers after multiple attempts. Reconnect device if this persists.")
```

I would try not to use prints here and rather logging to keep the stdout clean, logging allows a bit more control from the application side.

I think also throwing a proper exception if the modifiers are not there is better, then the applications would run uncalibrated, causing worse problems in the future... I have seen it before.

I also suggested (but didn't test) a way to hopefully harden the recovery process...

Tjoms99 replied:
I think, in hindsight, that the best way to solve this problem is to parse the metadata out from the garbage data (as it is in there). It feels like its the proper way to do it anyways, even on "clean" runs + it becomes more robust. Then, i think we can also remove the retries. 

If it fails, raising an IOError is a very good suggestion! 

I dont have time to look this before the weekend, but i would like to hear your thoughts on the suggestion above before continuing.

MrKevinWeiss replied:
I haven't closely looked into the output or the complexity of the parsing... If you think it is easy enough (and robust across the various PPK versions) then we can give it a try.  Solving hardware problems at the root is always better than adding retries and timeouts (though not always possible).

I am happy to test across firmware versions if you think you have it.  Otherwise, I have implemented this patch (or close enough to this patch) and am running tests with it.

MrKevinWeiss reviewed (changes requested):
A few small things, but at least now we can move!

Tjoms99 committed:
Empty buffer on ppk2 init, remove retries on metadata as it was rather treating the symptom instead of the root cause

Tjoms99 committed:
Add retries on connection as a bad exit can cause the serial port to not be immidiatly discovered

Tjoms99 committed:
Remove logging and raise exceptions instead. Calling logger.*(message) messes up with higher level application loggers. The application using the ppk2 lib should decide what to do with the exceptions and thereby log what they need.

Tjoms99 committed:
Fixed borken api call in power_profiler.py

Tjoms99 committed:
Harden API by adding types to functions

Tjoms99 commented:
Did another pass on this locally and verified it against a connected PPK2.

  A couple of extra things came up while testing/integrating this in other code using the API:

  - `get_modifiers()` is now tightened up so it either succeeds or raises, instead of returning a mixed `True` / `None`
  - added return type hints in `ppk2_api.py` for the main API methods so the return contracts are a bit clearer
  - fixed the compatibility issue in `src/power_profiler.py` where `get_samples()` was still treated like a flat list instead of `(samples, raw_digital)`

  I also removed the library-side logger usage that was ending up affecting my main application logger outside of ppk2. Hard failures now just raise instead of log-and-continue.

  Verified by running:
  `python3 examples/example_measure_mw.py`

  This still works on my side with a connected PPK2 (`/dev/ttyACM0`) and streams measurements as expected. How does it look on your end?

MrKevinWeiss commented on examples/example_measure_mw.py:
```diff
@@ -0,0 +1,107 @@
+import time
+import threading
+
+from src.ppk2_api.ppk2_api import PPK2_API
```
```suggestion
from ppk2_api.ppk2_api import PPK2_API
```

MrKevinWeiss replied:
Works with python3.10 (which is the oldest still supported) and I am a big fan of the type hints.

MrKevinWeiss reviewed (commented):
I tested with `python3.13` but needed a change...

<details><summary>manual testing results with "fresh" ppk</summary>

```
(.venv) weiss@lobaro-laptop-kw:~/repos/ppk2-api-python$ python examples/example_measure_mw.py
Press Enter to exit...
PPK2 devices found: [('/dev/ttyACM0', 'D06D19AF')]
[('/dev/ttyACM0', 'D06D19AF')]
Connecting to /dev/ttyACM0
Initialized Power Profiler
139.58 mW, Avg: 139.58 mW
27.09 mW, Avg: 83.34 mW
26.72 mW, Avg: 64.46 mW
26.66 mW, Avg: 55.01 mW
26.45 mW, Avg: 49.30 mW
26.29 mW, Avg: 45.47 mW
26.24 mW, Avg: 42.72 mW
26.17 mW, Avg: 40.65 mW
26.04 mW, Avg: 39.03 mW
26.02 mW, Avg: 37.73 mW
25.99 mW, Avg: 36.66 mW
25.95 mW, Avg: 35.77 mW
25.9 mW, Avg: 35.01 mW
25.83 mW, Avg: 34.35 mW
26.51 mW, Avg: 33.83 mW
26.94 mW, Avg: 33.40 mW
26.71 mW, Avg: 33.01 mW
27.01 mW, Avg: 32.67 mW
27.01 mW, Avg: 32.37 mW
26.99 mW, Avg: 32.11 mW
27.43 mW, Avg: 31.88 mW
27.78 mW, Avg: 31.70 mW
30.21 mW, Avg: 31.63 mW

28.85 mW, Avg: 31.52 mW
(.venv) weiss@lobaro-laptop-kw:~/repos/ppk2-api-python$ python examples/example_measure_mw.py
Press Enter to exit...
PPK2 devices found: [('/dev/ttyACM0', 'D06D19AF')]
[('/dev/ttyACM0', 'D06D19AF')]
Connecting to /dev/ttyACM0
Initialized Power Profiler
155.98 mW, Avg: 155.98 mW
26.94 mW, Avg: 91.46 mW
26.68 mW, Avg: 69.87 mW
26.61 mW, Avg: 59.05 mW
26.38 mW, Avg: 52.52 mW
26.22 mW, Avg: 48.13 mW
26.16 mW, Avg: 45.00 mW
26.09 mW, Avg: 42.63 mW
26.09 mW, Avg: 40.79 mW
25.98 mW, Avg: 39.31 mW
25.96 mW, Avg: 38.10 mW
25.92 mW, Avg: 37.08 mW
25.89 mW, Avg: 36.22 mW
25.84 mW, Avg: 35.48 mW
25.84 mW, Avg: 34.84 mW
25.82 mW, Avg: 34.27 mW
26.67 mW, Avg: 33.83 mW
26.96 mW, Avg: 33.45 mW
26.99 mW, Avg: 33.11 mW
27.0 mW, Avg: 32.80 mW
26.99 mW, Avg: 32.52 mW
27.01 mW, Avg: 32.27 mW
28.76 mW, Avg: 32.12 mW
27.48 mW, Avg: 31.93 mW
^C30.57 mW, Avg: 31.87 mW
Traceback (most recent call last):
  File "/home/weiss/repos/ppk2-api-python/examples/example_measure_mw.py", line 107, in <module>
    main()
    ~~~~^^
  File "/home/weiss/repos/ppk2-api-python/examples/example_measure_mw.py", line 102, in main
    input("Press Enter to exit...\n")
    ~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^
KeyboardInterrupt

(.venv) weiss@lobaro-laptop-kw:~/repos/ppk2-api-python$ python examples/example_measure_mw.py
Press Enter to exit...
PPK2 devices found: [('/dev/ttyACM0', 'D06D19AF')]
[('/dev/ttyACM0', 'D06D19AF')]
Connecting to /dev/ttyACM0
Initialized Power Profiler
131.61 mW, Avg: 131.61 mW
27.03 mW, Avg: 79.32 mW
26.74 mW, Avg: 61.79 mW
26.67 mW, Avg: 53.01 mW
26.45 mW, Avg: 47.70 mW
26.33 mW, Avg: 44.14 mW
26.21 mW, Avg: 41.58 mW
26.11 mW, Avg: 39.64 mW
26.07 mW, Avg: 38.14 mW
26.05 mW, Avg: 36.93 mW
26.0 mW, Avg: 35.93 mW
25.95 mW, Avg: 35.10 mW
25.9 mW, Avg: 34.39 mW
25.84 mW, Avg: 33.78 mW
26.46 mW, Avg: 33.29 mW
26.74 mW, Avg: 32.88 mW
26.8 mW, Avg: 32.53 mW
27.01 mW, Avg: 32.22 mW
26.99 mW, Avg: 31.95 mW
27.01 mW, Avg: 31.70 mW
27.43 mW, Avg: 31.50 mW
27.94 mW, Avg: 31.33 mW
29.77 mW, Avg: 31.27 mW

28.88 mW, Avg: 31.17 mW

```
</details>

MrKevinWeiss reviewed (approved):
The change was addressed, big improvement, thanks!
